simple-examples/FileIO.hs

```diff
@@ -0,0 +1,164 @@
+will induce a type error at compile time. That is, typechecking proves that all
+opened files are closed at some later point in execution. With the former
+interface, we have code that type checks but will error or cause errors at
+runtime.
+
+-}
+
+
+module FileIO where
+
+import Prelude
+import qualified System.IO as Sys
+import Control.Monad ()
+import Data.Text
+
+-- Linear Base Imports
+import qualified Control.Monad.Linear.Builder as Control
+import Data.Unrestricted.Linear
+import System.IO.Resource
+
+
+-- |  Non-linear first line printing
+--------------------------------------------
+
+-- openFile :: FilePath -> IOMode -> IO Handle
+-- IOMode = Sys.ReadMode | WriteMode | AppendMode | ReadWriteMode
+-- hGetLine :: Handle -> IO String
+-- hPutStr :: Handle -> String -> IO ()
+-- hClose :: Handle -> IO ()
+
+printFirstLine :: Sys.FilePath -> Sys.IO ()
+printFirstLine fpath = do
+  fileHandle <- Sys.openFile fpath Sys.ReadMode
+  firstLine <- Sys.hGetLine fileHandle
+  Sys.putStrLn firstLine
+  Sys.hClose fileHandle
+
+
+-- This compiles but can cause issues!
+printFirstLineNoClose :: Sys.FilePath -> Sys.IO ()
+printFirstLineNoClose fpath = do
+  fileHandle <- Sys.openFile fpath Sys.ReadMode
+  firstLine <- Sys.hGetLine fileHandle
+  Sys.putStrLn firstLine
+
+
+-- This compiles, but will error!
+printFirstLineAfterClose :: Sys.FilePath -> Sys.IO ()
+printFirstLineAfterClose fpath = do
+  fileHandle <- Sys.openFile fpath Sys.ReadMode
+  Sys.hClose fileHandle
+  firstLine <- Sys.hGetLine fileHandle
+  Sys.putStrLn firstLine
+
+
+-- | Linear first line printing
+--------------------------------------------
+
+linearGetFirstLine :: Sys.FilePath -> RIO (Unrestricted Text)
+linearGetFirstLine fp = do
+    handle <- openFile fp Sys.ReadMode
+    (t, handle') <- hGetLine handle
+    hClose handle'
+    return t
+  where
+    Control.Builder{..} = Control.monadBuilder
+
+linearPrintFirstLine :: Sys.FilePath -> Sys.IO ()
+linearPrintFirstLine fp = do
+  text <- run (linearGetFirstLine fp)
+  Sys.putStrLn (unpack text)
+
+
+{-
+    For clarity, we show this function without do notation.
+
+    Note that the current approach is limited.
+    We have to make the continuation use the unit type.
+
+    Enabling a more generic approach with a type index
+    for the multiplicity, as descibed in the paper is a work in progress.
+    This will hopefully result in using
+
+    `(>>==) RIO 'Many a #-> (a -> RIO p b) #-> RIO p b`
+
+    as the non-linear bind operation.
+-}
+
+-- | Linear and non-linear combinators
+-------------------------------------------------
+
+-- | Linear bind
+-- Notice the continuation has a linear arrow,
+-- i.e., (a #-> RIO b)
+(>>#=) :: RIO a #-> (a #-> RIO b) #-> RIO b
+(>>#=) = (>>=)
+  where
+    Control.Builder{..} = Control.monadBuilder
+
+-- | Non-linear bind
+-- Notice the continuation has a non-linear arrow,
+-- i.e., (() -> RIO b)
+(>>==) :: RIO () #-> (() -> RIO b) #-> RIO b
```

Comment:
A more general type would be

```suggestion
(>>==) :: RIO (Unrestricted a) #-> (a -> RIO b) #-> RIO b
```

You are not really making use of this function, though.

Reply:
Hmm ... yes, that's way better.

Reply:
Though, for the simplicity of this example, I'm not going to use that. (I'd have to fmap a `a #-> Unrestricted a` over `RIO ()` after using `hClose` and I want to keep things incredibly simple.